consensus/types/src/signed_beacon_block.rs

```diff
@@ -239,6 +240,45 @@ impl<E: EthSpec, Payload: AbstractExecPayload<E>> SignedBeaconBlock<E, Payload>
         }
     }
 
+    /// Produce a signed beacon block header AND a merkle proof for the KZG commitments.
+    ///
+    /// This method is more efficient than generating each part separately as it reuses hashing.
+    pub fn signed_block_header_and_kzg_commitments_proof(
+        &self,
+    ) -> Result<
+        (
+            SignedBeaconBlockHeader,
+            FixedVector<Hash256, E::KzgCommitmentsInclusionProofDepth>,
+        ),
+        Error,
+    > {
+        // Create the block body merkle tree
+        let body_leaves = self.message().body().body_merkle_leaves();
+        let beacon_block_body_depth = body_leaves.len().next_power_of_two().ilog2() as usize;
+        let body_merkle_tree = MerkleTree::create(&body_leaves, beacon_block_body_depth);
+
+        // Compute the KZG commitments inclusion proof
+        let (_, proof) = body_merkle_tree
+            .generate_proof(BLOB_KZG_COMMITMENTS_INDEX, beacon_block_body_depth)
+            .map_err(Error::MerkleTreeError)?;
+        let kzg_commitments_inclusion_proof = FixedVector::new(proof)?;
```

Comment:
I notice that this was duplicated from `kzg_commitments_merkle_proof` in the process of fixing the invalid proofs: https://github.com/sigp/lighthouse/pull/6268/commits/2f11f3cef33809124137bc5067a7fbdc957129fc

Is this because the last element of the proof _is not_ the body root, as this code had previously assumed? I see you're using `body_merkle_tree.hash()` to calculate it instead?

I don't quite understand what's going on here, as the comments on `MerkleTree::generate_proof`, suggest that the root of the tree _is_ the last element:

```rust
    /// Return the leaf at `index` and a Merkle proof of its inclusion.
    ///
    /// The Merkle proof is in "bottom-up" order, starting with a leaf node
    /// and moving up the tree. Its length will be exactly equal to `depth`.
```

What am I missing?

Reply:
Ok, I see the issue now. The merkle proof is not expected to contain the body root, as it is verified against the body root which is passed in as an additional argument in the [spec](https://github.com/ethereum/consensus-specs/blob/dev/specs/deneb/p2p-interface.md#verify_blob_sidecar_inclusion_proof):

```python
def verify_blob_sidecar_inclusion_proof(blob_sidecar: BlobSidecar) -> bool:
    gindex = get_subtree_index(get_generalized_index(BeaconBlockBody, 'blob_kzg_commitments', blob_sidecar.index))
    return is_valid_merkle_branch(
        leaf=blob_sidecar.kzg_commitment.hash_tree_root(),
        branch=blob_sidecar.kzg_commitment_inclusion_proof,
        depth=KZG_COMMITMENT_INCLUSION_PROOF_DEPTH,
        index=gindex,
        root=blob_sidecar.signed_block_header.message.body_root,
    )
```

I think the code duplication is OK for now.

Reply:
> Is this because the last element of the proof is not the body root, as this code had previously assumed?

Yes that's right. I also assumed the same and noticed that the body root was not returned from the `kzg_commitments_merkle_proof` function.

Here's how it looks like:

```
tree: Node(0xf260349e03f826c46f34efa8e943e0968294769a64b604d2504a80086561a2e0, Node(0x486d1c1a4f184afcdacc79f6db4d597d253b18cedc6650e5a0e476c0c6f77291, Node(0x6b2c3de18ef5ab5eab5b51436deaab9e030150aec6048801339ab285b45ae53f, Node(0x983ec82b9ba2e98e30cef7b0d0dcb14cb22acb45265996126bd53310a6e58a08, Leaf(0x8e742f5552fe005d58a4676d4c9709ac5cba3fc36cb99e387dcb2414edf53252), Leaf(0x0e07987a0f3b73e27a96321bedd723418f904225a74c471807203578b33e7217)), Node(0xea3336446bec447a262daa36ad1af5c3399d0ce139b6c923e687f5312d3a0800, Leaf(0x25748e4d40e0ebe492e451acf202dc39390cc685e22fa9a033341fff1bdb70a1), Leaf(0x8adfe28d95e5addbf1da4bff767f2693772471b83480cb6e4675b0983784a882))), Node(0xab16b427891c1851c565890e64dc70460d74166626f04a20a9088ea6ab232d93, Node(0x194b7b8ca7e899490d46bcdf67064131ce01db44e790f2f8a5c7c67dec4da2b4, Leaf(0x7a0501f5957bdf9cb3a8ff4966f02265f968658b7a9c62642cba1165e86642f5), Leaf(0x722f01799979caa2a3b3756f66e454c7e40e612cf6946e2f47b8012859798710)), Node(0x65740c64ece7e25b3d64491b394c082bef66cf91bd15fb0c135bca21c0025444, Leaf(0x792930bbd5baac43bcc798ee49aa8185ef76bb3b44ba62b91d86ae569e4bb535), Leaf(0x4a88fad606c5a81c01b68bfefbd93faba7e258046748da3a61975a3bb6c05212)))), Node(0x4ddafcd836f8a7ae8bb4c41159b4822054f7af664666882657dae73df482b33c, Node(0x3ffbaab784b1ede7fbbd558e6381cf6280bcc09f7b52cfa6321acfc240f3ec26, Node(0x5629b68a929f42c061c05145089d83b237783a93146326b25c171b97602d80e9, Leaf(0x08af26106a6b4c81ed49ee5080d465e42c6666e6163426116701922648c505cf), Leaf(0xdd63871f5155c221864d394560d89884201c2753ab67d833450a81a9d007aca7)), Node(0x877d453667ef92b0da105b398e24a2c59c104620075ae7b8b7bb11b85cbd0d93, Leaf(0x916a3b3c9be953a8db6d4887fee1dd2040ec57510fd8a747a1833fb0b806fd1b), Leaf(0xc52dde5d8ddefe2dc56a2a42484d964dbd146a4f8af2ee63aaecb240c2b4f79f))), Zero(2)))
proof: [0x916a3b3c9be953a8db6d4887fee1dd2040ec57510fd8a747a1833fb0b806fd1b, 0x5629b68a929f42c061c05145089d83b237783a93146326b25c171b97602d80e9, 0xdb56114e00fdd4c1f85c892bf35ac9a89289aaecb1ebd0a96cde606a748b5d71, 0x486d1c1a4f184afcdacc79f6db4d597d253b18cedc6650e5a0e476c0c6f77291]
body root: 0xf260349e03f826c46f34efa8e943e0968294769a64b604d2504a80086561a2e0
```

The function calculates the proof and discard the merkle tree, so the change here is just taking the `kzg_commitments_merkle_proof` logic here, and get the body root from the merkle tree, I believe there's no extra calculation.

I didn't change the existing function because it's used in multiple places.